ros/angel_system_nodes/angel_system_nodes/object_descriptor_generator.py

```diff
@@ -0,0 +1,323 @@
+        """
+        Callback function for the image subscriber. Performs image preprocessing,
+        model inference, output postprocessing, and detection set publishing for
+        each image received.
+        """
+        model = self.get_model()
+
+        # Preprocess image
+        im_in = np.array(BRIDGE.imgmsg_to_cv2(image, desired_encoding="rgb8"))
+        im_data, im_info, gt_boxes, num_boxes, im_scales = self.preprocess_image(im_in)
+
+        # Send to model
+        rois, cls_prob, \
+        _, _, _, _, _, _, \
+        pooled_feat = model(im_data, im_info, gt_boxes, num_boxes, pool_feat=True)
+
+        # Postprocess model output
+        detection_info = self.postprocess_detections(
+            rois, cls_prob,
+            pooled_feat, im_scales
+        )
+
+        # Form into object detection set message
+        msg = ObjectDetection2dSet()
+        msg.header.stamp = self.get_clock().now().to_msg()
+        msg.header.frame_id = image.header.frame_id
+        msg.source_stamp = image.header.stamp
+
+        if detection_info['boxes'] is None:
+            msg.num_detections = 0
+        else:
+            msg.num_detections = len(detection_info['labels'])
+            msg.label_vec = self.classes[1:]
+            msg.label_confidences = detection_info['scores'].ravel().tolist()
+
+            msg.left = detection_info['boxes'][:,0].tolist()
+            msg.top = detection_info['boxes'][:,1].tolist()
+            msg.right = detection_info['boxes'][:,2].tolist()
+            msg.bottom = detection_info['boxes'][:,3].tolist()
+
+            msg.descriptor_dims = list(detection_info['feats'].shape)
+            msg.descriptors = detection_info['feats'].ravel().tolist()
+
+        # Publish detection set message
+        self._publisher.publish(msg)
+
+    def preprocess_image(self, im_in):
+        """
+        Preprocess the image and return the necessary inputs for the fasterrcnn
+        mode. Returns the image blob data, image info, ground truth boxes, image
+        scales, and number of boxes.
+
+        Based on:
+        https://github.com/shilrley6/Faster-R-CNN-with-model-pretrained-on-Visual-Genome/blob/master/generate_tsv.py
+        """
+        # initilize the tensor holder here.
+        im_data = torch.FloatTensor(1).to(device=self._torch_device)
+        im_info = torch.FloatTensor(1).to(device=self._torch_device)
+        num_boxes = torch.LongTensor(1).to(device=self._torch_device)
+        gt_boxes = torch.FloatTensor(1).to(device=self._torch_device)
+
+        # Make variable
+        with torch.no_grad():
+            im_data = Variable(im_data)
+            im_info = Variable(im_info)
+            num_boxes = Variable(num_boxes)
+            gt_boxes = Variable(gt_boxes)
+
+        if len(im_in.shape) == 2:
+            im_in = im_in[:,:,np.newaxis]
+            im_in = np.concatenate((im_in,im_in,im_in), axis=2)
+        im = im_in
+
+        blobs, im_scales = _get_image_blob(im)
```

Comment:
`_get_image_blob` is documented saying that it expects a "BGR" image, but I expect that at this point the image input is in RGB (see L125).

Reply:
Good catch!